### PR TITLE
Remove the implementation detail leak, implemented in swift

### PR DESCRIPTION
**Pull request by drkibitz**
Hello thank you for this great work!

This is my attempt at move "leaky details" in the public header to Swift.
My motivations are writing a pure swift package via Swift Package Manager.
You will also notice that the git2.h header no longer needs to be public either.

I also have a followup the last .m item (the constructor), to a central class, calling git_libgit2_init on init and shutdown on deinit, like this: https://github.com/drkibitz/swift-Git2/blob/master/Sources/Git2/Git2.swift

### CODE REVIEW

File: SwiftGit2.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -74,7 +74,7 @@
 		621E66A71C72958800A0F352 /* Pointers.swift in Sources */ = {isa = PBXBuildFile; fileRef = BE7A753E1A4A2BCC002DA7E3 /* Pointers.swift */; };
 		621E66A81C72958800A0F352 /* Errors.swift in Sources */ = {isa = PBXBuildFile; fileRef = DA5914751A94579000AED74C /* Errors.swift */; };
 		621E66A91C72958800A0F352 /* SwiftGit2.m in Sources */ = {isa = PBXBuildFile; fileRef = BE14AA4F1A1974010015B439 /* SwiftGit2.m */; };
-		621E66AF1C72958800A0F352 /* git2.h in Headers */ = {isa = PBXBuildFile; fileRef = BEB31F9B1A0E581400F525B9 /* git2.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (mdiep) [inline SwiftGit2.xcodeproj/project.pbxproj]: Could you do this is a separate PR? I'm uncertain about this part, and I don't want it to hold up the rest of the changes.

File: SwiftGit2/Repository.swift [hunk 2]
```diff
@@ -9,7 +9,32 @@
...
+func SG2CheckoutProgressCallback(path: UnsafePointer<Int8>, completed_steps: Int, total_steps: Int, payload: UnsafeMutablePointer<Void>) -> Void {
+	if (payload != nil) {
+		let buffer = UnsafeMutablePointer<CheckoutProgressBlock>(payload)
+		let block: CheckoutProgressBlock = buffer.memory as CheckoutProgressBlock
+		block((path == nil ? nil : String(path)), completed_steps, total_steps);
+	}
+}
+
+func SG2CheckoutOptions(progress: CheckoutProgressBlock? = nil) -> git_checkout_options {
+
```
Comment (mdiep) [inline SwiftGit2/Repository.swift]: Could you remove this empty line?

File: SwiftGit2/Repository.swift [hunk 3]
```diff
@@ -9,7 +9,32 @@
 import Foundation
 import Result
 
-public typealias CheckoutProgressBlock = SG2CheckoutProgressBlock
+public typealias CheckoutProgressBlock = (String?, Int, Int) -> Void
+
+func SG2CheckoutProgressCallback(path: UnsafePointer<Int8>, completed_steps: Int, total_steps: Int, payload: UnsafeMutablePointer<Void>) -> Void {
+	if (payload != nil) {
+		let buffer = UnsafeMutablePointer<CheckoutProgressBlock>(payload)
+		let block: CheckoutProgressBlock = buffer.memory as CheckoutProgressBlock
```
Comment (mdiep) [inline SwiftGit2/Repository.swift]: We should only need to specify the type once on this line

File: SwiftGit2/Repository.swift [hunk 4]
```diff
@@ -9,7 +9,32 @@
...
+func SG2CheckoutOptions(progress: CheckoutProgressBlock? = nil) -> git_checkout_options {
+
+	// Do this because GIT_CHECKOUT_OPTIONS_INIT is unavailable in swift
+	let pointer = UnsafeMutablePointer<git_checkout_options>.alloc(1)
+	git_checkout_init_options(pointer, UInt32(GIT_CHECKOUT_OPTIONS_VERSION))
+
+	var result: git_checkout_options = pointer.memory
+	if (progress != nil) {
+		result.progress_cb = SG2CheckoutProgressCallback
+		let aProgress = UnsafeMutablePointer<CheckoutProgressBlock>.alloc(1)
```
Comment (mdiep) [inline SwiftGit2/Repository.swift]: If we call `alloc`, we also need to call `dealloc`.

**Comment by mdiep** [comment]
Thanks for the pull request! I'm really excited to move this code to Swift. When I originally wrote the code, that wasn't possible.

**Comment by drkibitz** [comment]
@mdiep thanks for the comments, new commits here with :cake:

**Comment by drkibitz** [comment]
Sorry got a little trigger happy ;)

File: SwiftGit2/Repository.swift [hunk 5]
```diff
@@ -9,7 +9,34 @@
...
 import Result
 
-public typealias CheckoutProgressBlock = SG2CheckoutProgressBlock
+public typealias CheckoutProgressBlock = (String?, Int, Int) -> Void
+
+func SG2CheckoutProgressCallback(path: UnsafePointer<Int8>, completed_steps: Int, total_steps: Int, payload: UnsafeMutablePointer<Void>) -> Void {
+	if (payload != nil) {
+		let buffer = UnsafeMutablePointer<CheckoutProgressBlock>(payload)
+		let block = buffer.move()
+		payload.dealloc(1)
```
Comment (mdiep) [inline SwiftGit2/Repository.swift]: This doesn't seem right. This block should be called multiple times in a checkout.

Comment (drkibitz) [inline SwiftGit2/Repository.swift]: Right, was thinking of a it as a completion for some reason. Also something to note is that this doesn't happen in the test.

**Comment by drkibitz** [comment]
Updated, but... I'm not really sure, should I move the alloc for the callback pointer to the checkout method, and dealloc there too?

**Comment by drkibitz** [comment]
@mdiep Pretty sure I just I answered my own question... just check for completion in the progress callback and dealloc then. Also, I'm going to squash these commits, and push it back up since this PR is getting a git ugly.

**Comment by mdiep** [comment]
Looks good—thanks!

(Sorry it took me so long to review this!)
